### PR TITLE
ci: add sccache to speed up ci build

### PR DESCRIPTION
**Pull request by xxchan**
e.g., the `unit` job takes 10min in total, and build takes 5min. So I think we can introduce cache to speed it up.
 
image

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -104,6 +106,9 @@ jobs:
     steps:
       - uses: actions/checkout@v4
 
+      - name: Setup sccache
+        uses: mozilla-actions/sccache-action@v0.0.4
```
Comment (xxchan) [inline .github/workflows/ci.yml:110]: The latest version is 0.0.7, but I only see 0.0.4 listed here https://github.com/apache/infrastructure-actions/blob/main/approved_patterns.yml

**Comment by xxchan** [comment]
comparison:
before: https://github.com/apache/iceberg-rust/actions/runs/12394303990
image

after: https://github.com/apache/iceberg-rust/actions/runs/12395899529?pr=824

image

Seems not as good as expected, but generally reduced >1 min. UT reduced >2 min

**Comment by xxchan** [comment]
Seems caused by https://github.com/Mozilla-Actions/sccache-action/issues/50

**Comment by Fokko** [comment]
Windows build took even longer after the change 🤔

**Comment by xxchan** [comment]
The cache hit rate was very low, and there's plenty of Cache write errors (I guess it's rate limit https://github.com/Mozilla-Actions/sccache-action/issues/50).

After running it for more times to warm up the cache, it should perform better.

image

**Comment by xxchan** [comment]
I found an alternative to `sccache`: `swatinem/rust-cache@v2` (also approved by ASF)

They work differently:
- `sccache` writes/reads cache for each individual build artifact.
- `rust-cache` has a large cache for all dependencies (keyed by `Cargo.toml` and `Cargo.lock`) 

It's possible that `rust-cache` works better with GHA cache. But we will need to check in `Cargo.lock` (currently ignored) to make it work better. (We can consider doing it, and it has other benefits like a reproduciable build for CI and all developers.)

File: .github/workflows/ci.yml (hunk 1 above)
Comment (kevinjqliu) [reply .github/workflows/ci.yml:120]: heres the original issue for 0.0.4

https://issues.apache.org/jira/browse/INFRA-25525?jql=text%20~%20%22sccache-action%22

**Comment by xxchan** [comment]
To be more precise about the performance gain: For unit test: previous compile time is `5m 06s`, now `2m 35s`

cache stats:

image

**Comment by xxchan** [comment]
> I'm not a big fan of checking in Cargo.lock as it's an antipattern for library

Just FYI that it's not considered an antipattern any more https://blog.rust-lang.org/2023/08/29/committing-lockfiles.html

And e.g., opendal has it: https://github.com/apache/opendal/blob/main/core/Cargo.lock

I'm OK with both though.

**Comment by kevinjqliu** [comment]
Since this PR only affects CI and seems to be an overall improvement. I don't see any downside of including this

**Comment by Fokko** [comment]
The downside is additional complexity and maintenance. The Windows job takes the most time, and after the change it takes even more, so the maximum runtime gets worse, just the average runtime improves. I'm leaning toward what @liurenjie1024 is saying, and avoiding caching. Curious to learn what @Xuanwo thinks

**Comment by xxchan** [comment]
replaced by https://github.com/apache/iceberg-rust/pull/844

**Pull request closed without merging**